Sources/SuperwallKit/Network/Network.swift

```diff
@@ -0,0 +1,198 @@
+  private var applicationStatePublisher: AnyPublisher<UIApplication.State, Never> {
+    UIApplication.shared.publisher(for: \.applicationState)
+      .eraseToAnyPublisher()
+  }
+  private let factory: ApiFactory
+
+  init(
+    urlSession: CustomURLSession = CustomURLSession(),
+    factory: ApiFactory
+  ) {
+    self.urlSession = urlSession
+    self.factory = factory
+  }
+
+  func sendEvents(events: EventsRequest) async {
+    do {
+      let result = try await urlSession.request(.events(eventsRequest: events, factory: factory))
+      switch result.status {
+      case .ok:
+        break
+      case .partialSuccess:
+        Logger.debug(
+          logLevel: .warn,
+          scope: .network,
+          message: "Request had partial success: /events",
+          info: ["payload": result.invalidIndexes as Any]
+        )
+      }
+    } catch {
+      Logger.debug(
+        logLevel: .error,
+        scope: .network,
+        message: "Request Failed: /events",
+        info: ["payload": events],
+        error: error
+      )
+    }
+  }
+
+  func getPaywall(
+    withId identifier: String? = nil,
+    fromEvent event: EventData? = nil
+  ) async throws -> Paywall {
+    do {
+      return try await urlSession.request(
+        .paywall(withIdentifier: identifier, fromEvent: event, factory: factory)
+      )
+    } catch {
+      if identifier == nil {
+        Logger.debug(
+          logLevel: .error,
+          scope: .network,
+          message: "Request Failed: /paywall",
+          info: [
+            "identifier": identifier ?? "none",
+            "event": event.debugDescription
+          ],
+          error: error
+        )
+      } else {
+        Logger.debug(
+          logLevel: .error,
+          scope: .network,
+          message: "Request Failed: /paywall/:identifier",
+          error: error
+        )
+      }
+      throw error
+    }
+  }
+
+  func getPaywalls() async throws -> [Paywall] {
+    do {
+      let response = try await urlSession.request(
+        .paywalls(factory: factory),
+        isForDebugging: true
+      )
+      return response.paywalls
+    } catch {
+      Logger.debug(
+        logLevel: .error,
+        scope: .network,
+        message: "Request Failed: /paywalls",
+        error: error
+      )
+      throw error
+    }
+  }
+
+  func getConfig(
+    withRequestId requestId: String,
+    injectedApplicationStatePublisher: (AnyPublisher<UIApplication.State, Never>)? = nil
+  ) async throws -> Config {
+    // Suspend until app is in foreground.
+    let applicationStatePublisher = injectedApplicationStatePublisher ?? self.applicationStatePublisher
+
+    await applicationStatePublisher
+      .subscribe(on: DispatchQueue.main)
+      .filter { $0 != .background }
+      .eraseToAnyPublisher()
+      .async()
+
+    do {
+      var config = try await urlSession.request(.config(requestId: requestId, factory: factory))
+      config.requestId = requestId
```

Comment:
Setting the `requestId` here seemed off to me, given the `urlSession.request(...)` knows what the `requestId` is – why does the caller need to set it? And in what cases (e.g. `confirmAssignments()` below also calls `urlSession.request` but doesn't set the `requestId`)

Reply:
This was for TriggerSessions (which aren't actually used now, but we're keeping them in anyway for now...). We needed to have the config request ID for some reason. I think the issue is that the actual request function is generic. So either I put this code in that generic function:

```
if var value = value as? Config {
  value.requestId = requestId
}
```
Which seems like a waste as it'll get evaluated every request, or I put it in the `getConfig` function. There's no reason why the UUID can't be generated within getConfig, but not sure about moving it into `request()`. Seems like a downside to using generics